kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -490,17 +501,33 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
                         }
                         if (rootCause != null) {
                             // Note: attachChild uses invokeImmediately, so it gets invoked when adding to cancelled job
-                            if (invokeImmediately) handler.invokeIt(rootCause)
+                            if (invokeImmediately) {
+                                if (nodeAdded != null) {
+                                    nodeAdded!!.invoke(rootCause)
+                                } else {
+                                    handler.invokeIt(rootCause)
+                                }
+                            }
                             return handle
                         } else {
-                            if (addLastAtomic(state, list, node)) return node
+                            // if ChildCompletion is already invoked, return handler, so that parent would wait for the child to complete
+                            nodeAdded = node
+                            if (addLast(state, list, node) || node is ChildCompletion && !node.markInvoked()) {
+                                return node
+                            }
                         }
                     }
                 }
                 else -> { // is complete
                     // :KLUDGE: We have to invoke a handler in platform-specific way via `invokeIt` extension,
                     // because we play type tricks on Kotlin/JS and handler is not necessarily a function there
-                    if (invokeImmediately) handler.invokeIt((state as? CompletedExceptionally)?.cause)
+                    if (invokeImmediately) {
+                        if (nodeAdded != null) {
```

Comment:
Seems like it can be replaced with `(nodeAdded ?: node).invoke((state as? CompletedExceptionally)?.cause)`